arch/x86/kvm/pkvm.c

```diff
@@ -105,6 +106,125 @@ static int enable_vmexit_trace(void *data, u64 val)
 }
 DEFINE_SIMPLE_ATTRIBUTE(enable_vmexit_trace_fops, NULL, enable_vmexit_trace, "%llu\n");
 
+static const struct vmexit_reason {
+	int reason;
+	char *name;
+} reasons[] = { VMX_EXIT_REASONS, { -1, NULL } };
+
+static const char *get_vmexit_reason(int reason)
+{
+	const struct vmexit_reason *p = reasons;
+
+	while (p->name) {
+		if (p->reason == reason)
+			return p->name;
+		p++;
+	}
+
+	return NULL;
+}
+
+static int print_perf_data(struct seq_file *m, struct perf_data *summary,
+			   struct perf_data *pervcpu)
+{
+	struct perf_data *print = pervcpu ? pervcpu : summary;
+	struct vmexit_stats *stats, *summary_stats;
+	const char *prefix;
+	int *vcpu_id;
+	int i;
+
+	if (!print)
+		return -EINVAL;
+
+	vcpu_id = (print == pervcpu) ? &pervcpu->vcpu_id : NULL;
+	if (vcpu_id)
+		prefix = kasprintf(GFP_KERNEL, "Host-vcpu%d: ", *vcpu_id);
+	else
+		prefix = kasprintf(GFP_KERNEL, "Host: ");
+
+	if (!prefix)
+		return -ENOMEM;
+
+	for (summary_stats = (print == pervcpu && summary) ? summary->vmexit_reasons : NULL,
+	     stats = print->vmexit_reasons, i = 0;
+	     i < MAX_EXIT_REASONS;
+	     summary_stats = summary_stats ? summary_stats + 1 : NULL, stats++, i++) {
+		if (!stats->count)
+			continue;
+
+		seq_printf(m, "%s%s %lld cycles %lld each-handler-cycle %lld\n",
+			   prefix, get_vmexit_reason(i), stats->count, stats->cycles,
+			   stats->cycles / stats->count);
+
+		if (summary_stats) {
+			summary_stats->count += stats->count;
+			summary_stats->cycles += stats->cycles;
+		}
+
+		if (need_resched())
+			cond_resched();
+	}
+
+	kfree(prefix);
+
+	return 0;
+}
+
+static int pkvm_dump_vmexit_trace(struct seq_file *m, struct perf_data *dump,
+				  unsigned long size)
+{
+	struct perf_data *summary = kmalloc(sizeof(struct perf_data), GFP_KERNEL_ACCOUNT);
+	struct perf_data *pervcpu = dump;
```

Comment:
nit: for consistency we might do `for (cpu = 0, pervcpu = dump;` in the loop below, instead of initializing it here?